src/Type/Doctrine/Query/QueryType.php

```diff
@@ -14,10 +14,15 @@ class QueryType extends GenericObjectType
 	/** @var string */
 	private $dql;
 
-	public function __construct(string $dql, ?Type $resultType = null)
-	{
-		$resultType = $resultType ?? new MixedType();
-		parent::__construct('Doctrine\ORM\Query', [$resultType]);
+	public function __construct(
+		string $dql,
+		?Type $resultType,
+		?Type $indexType
+	) {
+		parent::__construct('Doctrine\ORM\Query', [
+			$resultType ?? new MixedType(),
+			$indexType ?? new MixedType(),
+		]);
```

Comment:
Right now we're incompatible with Psalm https://github.com/phpstan/phpstan-doctrine/issues/377 so I wouldn't avoid the change now.

Reply:
What about using some artificial custom type for this purpose? Like `ListIndexMarkerType extends MixedType`?

Reply:
I think I fixed this issue, hopefully. 

Default value of `QueryType` is still `MixedType`. But only when we get valid answer from `QueryResultTypeBuilder::getIndexType`, we get either `ListIndexMarkerType` or the type of indexBy type (can be mixed). But the extension now infers lists only for `ListIndexMarkerType` so we should be getting `list<T>` only in those cases where it really passed the `QueryResultTypeWalker` without any issues.

Reply:
If we want to be compatible with Psalm, we need to put the TIndex param before TResult as well.

The artificially custom type is a good idea. Will the user be able to type a list Query in type annotations like in the following snippet ?

``` php
/** @return Query<ListQuery,Foo> */
function getQuery(): Query;
```

Reply:
We can change the types only for bleedingEdge. We're doing this whole PR for the list type which is only in bleedingEdge anyway.

Reply:
That seems hard. Wouldnt it be actually better to sync with psalm even in non-bleedingEdge? Meaning perform this (imo minor) BC-break. We would need to do it some time in future anyway.

We use phpstan-doctrine infering quite lot in our codebase, but we never needed to use the generics ourselves. Do you @arnaud-lb use it in userland?

Reply:
Note: I just used this branch as-is in our codebase without any issues and it allowed me to actually enable listType feature.

Reply:
> Maybe we could introduce a subtype of `Query<TResult>`, like `IndexedQuery<TIndex,TResult>` ?

I'm withdrawing this suggestion, it wouldn't work.

After all we should probably add `TKey`, as we will have to do this change if we add full INDEX BY support anyway, and this makes us compatible with Psalm at the same time.

Users will have to change `Query<Something>` to `Query<int,Something>` for non-indexed queries or `Query<mixed,Something>` for indexed queries (and to something more specific than `mixed` if we add full INDEX BY support).

> We use phpstan-doctrine infering quite lot in our codebase, but we never needed to use the generics ourselves. Do you @arnaud-lb use it in userland?

I used the `Query<...>` or `AbstractQuery<...>` types on only a few occasions, but it would have been very inconvenient if the result of some `createQuery()` call couldn't be accepted by any `Query<...>` type. It would be good enough if `Query<mixed,Something>` or `Query<int,Something>` accepts a QueryType that have TKey=ListIndexMarkerType (I don't know if it does currently, we may need to declare TKey as @template-covariant, which would be correct).

PHPDoc types are resolved in [TypeNodeResolver](https://github.com/phpstan/phpstan-src/blob/d0207ed79e94e4b4daceb0f83abb1f481ee3250b/src/PhpDoc/TypeNodeResolver.php), and plugins can implement TypeNodeResolverExtensions to add custom types: https://phpstan.org/developing-extensions/custom-phpdoc-types

Alternatively, we could bind TKey to the `void` type to denote a non-indexed Query. We bind TResult to `void` to denote non-SELECT queries already.

Reply:
> PHPDoc types are resolved in [TypeNodeResolver](https://github.com/phpstan/phpstan-src/blob/d0207ed79e94e4b4daceb0f83abb1f481ee3250b/src/PhpDoc/TypeNodeResolver.php), and plugins can implement TypeNodeResolverExtensions to add custom types: https://phpstan.org/developing-extensions/custom-phpdoc-types

I can add support of some `list-marker` phpdoc for the `ListIndexMarkerType`, but I think it might be confusing for people seeing it error messages. But the truth is that this "hidden" type is not ideal as well (it is outputted as any other `mixed` now).

Reply:
I'd love to finalize this. I'm just waiting for your answers / review. Thanks.